Basic vfio-user server support

rbradford opened the pull request:
- vfio_user: Improve flags for DMA_MAP command
- vfio_user: Add flags for DMA_UNMAP command
- vfio_user: Add basic server side support
- vfio_user: Add a simple sample device

rbradford committed:
vfio_user: Improve flags for DMA_MAP command

Replace the use of an enum with a bitfield representation which means
that is now possible to logical OR flags together.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Add flags for DMA_UNMAP command

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Add basic server side support

This allows the implementation of PCI devices in a different process
using the vfio-user protocol.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Add a simple sample device

This implements a similar GPIO device to that found in the libvfio-user
source code.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

likebreath reviewed (commented):
PR looks good with the first pass. I will take a closer look next Monday.

It may worth expanding the current vfio-user documentation with contents on how to use the added gpio example as a vfio-user device with Cloud Hypervisor?

sameo reviewed (approved):
Thanks a lot @rbradford

likebreath commented on vfio_user/src/lib.rs:
```diff
@@ -782,3 +791,482 @@ impl Client {
+                .map(|fd| unsafe { File::from_raw_fd(*fd) })
+                .collect();
+
+            match header.command {
```
I think we should always send an `Error` reply to the client if the server stops running due to errors. One idea is we can put this big `match` block into a `catch_unwind` statement, form which we always catch any errors and inform the client before stop running the server. Also, it would simply the `match` block by avoiding sending "Error" replies individually from some of the commands.

rbradford replied:
Great suggestion!

likebreath reviewed (commented):
Only one proposal about the general design: I think it is better that we always inform the client with an "Error" reply when the server stops running due to errors. 

Also some specific questions and comments below (mostly nits or questions from vfio user spec).

rbradford committed:
vfio_user: Always generate an error for commands

Any error from the backend or from the protocol handling code will now
result in an error reply being sent. This is cleanly achieved by
splitting the command handling out into its own method and using the
Rust Result<> based error handling to trigger the generation of the
error reply.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford commented on vfio_user/src/lib.rs:
```diff
@@ -782,3 +791,482 @@ impl Client {
+                        client_version.major, client_version.minor, client_capabilities
+                    );
+
+                    let version = Version {
+                        header: Header {
+                            message_id: client_version.header.message_id,
+                            command: Command::Version,
+                            flags: HeaderFlags::Reply as u32,
+                            message_size: (size_of::<Version>() + version_data.len() + 1) as u32,
+                            ..Default::default()
+                        },
+                        major: 0,
+                        minor: 1,
+                    };
+
+                    let server_capabilities = Capabilities::default();
```
I don't think it should be configurable right now  as the protocol is still unstable but I have added a TODO.

rbradford committed:
vfio_user: Add TODOs for missing functionality

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Reject SET_IRQS with VFIO_IRQ_SET_DATA_BOOL

This is unsupported.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Add command line parsing for socket option

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
vfio_user: Make GPIO device more interesting

Every third read on the GPIO pin will return 1 and also trigger an
interrupt in the guest.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford commented:
I improved this PR a bit; the GPIO is now more interesting in that it returns different values based on how many times it has been read as well as firing an interrupt.

likebreath commented on vfio_user/Cargo.toml:
```diff
@@ -13,4 +17,9 @@ serde_json = "1.0.93"
 thiserror = "1.0.38"
 vfio-bindings = { git = "https://github.com/rust-vmm/vfio", branch = "main", features = ["fam-wrappers"] }
 vm-memory = { version = "0.10.0", features = ["backend-mmap", "backend-atomic"] }
```
To build the gpio example (with `cargo build -p vfio_user --example gpio`, I will need to enable the `"backend-bitmap"` feature for the `vm-memory` crate, otherwise I will get the following error

```
error[E0432]: unresolved import `vm_memory::bitmap::AtomicBitmap`                                                                                                                                            
  --> pci/src/vfio_user.rs:23:5                                                                                                                                                                               
   |                                                                                                                                                                                                          
23 | use vm_memory::bitmap::AtomicBitmap;                                                                                                                                                                     
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `AtomicBitmap` in `bitmap`                                                                                                                                        
                                                                                                                                                                                                              
For more information about this error, try `rustc --explain E0432`
```

rbradford replied:
Okay, so I didn't see that because I used `cargo build --all --examples` to build it. The correct fix was to ensure the pci crate used the correct features for the vm-memory crate.

rbradford replied:
I did enhance the GH workflow to ensure buildability!

likebreath commented:
> I improved this PR a bit; the GPIO is now more interesting in that it returns different values based on how many times it has been read as well as firing an interrupt.

I tried the gpio device and the writing pattern works. I encountered an error related to the `irq`. If I hotplug and then remove the gpio device from the VM via `ch-remote`, the gpio device would panic with the following error:

```
$ ./gpio --socket-path /tmp/vu-socket
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', vfio_user/examples/gpio/main.rs:160:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

At the same time, the CH process would generate an error message:
```
cloud-hypervisor: 293.609211s: <vcpu0> ERROR:pci/src/vfio.rs:904 -- Could not disable INTx: VFIO user error: Error reading from stream: failed to fill whole buffer
```

rbradford committed:
build: Ensure pci crate uses vm-memory with required features

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
github: Build examples as well as tests in quality workflow

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford commented:
> > I improved this PR a bit; the GPIO is now more interesting in that it returns different values based on how many times it has been read as well as firing an interrupt.
> 
> I tried the gpio device and the writing pattern works. I encountered an error related to the `irq`. If I hotplug and then remove the gpio device from the VM via `ch-remote`, the gpio device would panic with the following error:
> 
> ```
> $ ./gpio --socket-path /tmp/vu-socket
> thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', vfio_user/examples/gpio/main.rs:160:29
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```
> 
> At the same time, the CH process would generate an error message:
> 
> ```
> cloud-hypervisor: 293.609211s: <vcpu0> ERROR:pci/src/vfio.rs:904 -- Could not disable INTx: VFIO user error: Error reading from stream: failed to fill whole buffer
> ```

Cool - thanks for trying this, I got hotplug/unplug working now.
